providers/attio/metadata.go

```diff
@@ -0,0 +1,118 @@
+// nolint
+package attio
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/amp-labs/connectors/common"
+	"github.com/spyzhov/ajson"
+)
+
+// This struct is used for when the response data having slice of data.
+type responseObject struct {
+	Result []map[string]any `json:"data"`
+	// Other fields
+}
+
+var errCannotLoadMetadata = errors.New("cannot load metadata")
+
+// ListObjectMetadata creates metadata of object via reading objects using Attio API.
+func (c *Connector) ListObjectMetadata(ctx context.Context,
+	objectNames []string,
+) (*common.ListObjectMetadataResult, error) {
+	if len(objectNames) == 0 {
+		return nil, common.ErrMissingObjects
+	}
+
+	metadataResult := common.ListObjectMetadataResult{
+		Result: make(map[string]common.ObjectMetadata),
+		Errors: make(map[string]error),
+	}
+
+	for _, obj := range objectNames {
+		// Constructing the request url.
+		url, err := c.getApiURL(obj)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := c.Client.Get(ctx, url.String())
+		if err != nil {
+			metadataResult.Errors[obj] = err
+
+			continue
+		}
+
+		var metadata *common.ObjectMetadata
+		if obj == "self" {
+			// Getting the metadata for self object in separate function
+			metadata, err = parseMetadataForSelfObject(resp)
+		} else {
+			metadata, err = parseMetadataFromResponse(resp)
+		}
+
+		if err != nil {
+			if errors.Is(err, errCannotLoadMetadata) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+
+		metadata.DisplayName = obj
+		metadataResult.Result[obj] = *metadata
+	}
+
+	return &metadataResult, nil
+}
+
+func parseMetadataFromResponse(resp *common.JSONHTTPResponse) (*common.ObjectMetadata, error) {
+	response, err := common.UnmarshalJSON[responseObject](resp)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(response.Result) == 0 {
+		return nil, errCannotLoadMetadata
+	}
+
+	metadata := &common.ObjectMetadata{
+		FieldsMap: make(map[string]string),
+	}
+
+	// Using the first result data to generate the metadata.
+	for k := range response.Result[0] {
+		metadata.FieldsMap[k] = k
+	}
+
+	return metadata, nil
+}
+
+func parseMetadataForSelfObject(resp *common.JSONHTTPResponse) (*common.ObjectMetadata, error) {
+	body, ok := resp.Body()
+	if !ok {
+		return nil, errCannotLoadMetadata
+	}
+
+	bodyBytes, err := ajson.Marshal(body)
+	if err != nil || len(bodyBytes) == 0 {
+		return nil, errCannotLoadMetadata
+	}
+
+	// Parse the JSON bytes into a map[string]interface{}
+	var responseMap map[string]interface{}
+	if err := json.Unmarshal(bodyBytes, &responseMap); err != nil {
+		return nil, err
+	}
+
+	fieldsMap := map[string]string{}
+	for k := range responseMap {
```

Comment:
Since we are only collecting "keys", `ajson` library has method that will avoid additional marshall, unmarshall. It would look like this:
```
responseMap, err := body.GetObject()
```
https://github.com/spyzhov/ajson/blob/0d5fdeed023cf269f55ec781a2713153e3f3ca33/node.go#L494


Alternatively, if you want to deal with byte data directly avoiding ajson library, you could use the same UnmarshalJSON you use above but for map type:
```
responseMap, err := common.UnmarshalJSON[map[string]any](resp)
```

Reply:
You are serving Metadata from API and so far we were adding static `schema.json` for those providers that do not have such APIs. Therefore, no action is needed on this one.

.

**PS**: you can check out pipeliner to see how `schema.json` is created and served.
There is OpenAPI yaml or json commited under:
https://github.com/amp-labs/connectors/tree/main/providers/pipeliner/openapi
The `FileManager` loads it (file.go) and is referenced by 2 places:
* Schemas are objects object representation of that file and is used by `ListObjectMetadata` https://github.com/amp-labs/connectors/blob/main/providers/pipeliner/metadata.go.
* And most importantly this is the script which will differ among providers: https://github.com/amp-labs/connectors/blob/main/scripts/openapi/pipeliner/metadata/main.go. It loads OpenAPI file, processes it in custom way and outputs the `schema.json`.

In rare cases `schema.json` is created manually or a website is scrapped, which should be avoided for maintenance cost: https://github.com/amp-labs/connectors/tree/main/scripts/scrapper/salesloft/metadata

Here are some READMEs that I wrote:
* https://github.com/amp-labs/connectors/tree/main/scripts/openapi
* https://github.com/amp-labs/connectors/tree/main/tools/fileconv/api3